ic-identity-hsm/src/hsm.rs

```diff
@@ -0,0 +1,358 @@
+use ic_agent::{Identity, Signature};
+use ic_types::Principal;
+
+use num_bigint::BigUint;
+use openssl::sha::Sha256;
+use pkcs11::types::{
+    CKA_CLASS, CKA_EC_PARAMS, CKA_EC_POINT, CKA_ID, CKA_KEY_TYPE, CKF_LOGIN_REQUIRED,
+    CKF_SERIAL_SESSION, CKK_ECDSA, CKM_ECDSA, CKO_PRIVATE_KEY, CKO_PUBLIC_KEY, CKU_USER,
+    CK_ATTRIBUTE, CK_ATTRIBUTE_TYPE, CK_KEY_TYPE, CK_MECHANISM, CK_OBJECT_CLASS, CK_OBJECT_HANDLE,
+    CK_SESSION_HANDLE, CK_SLOT_ID,
+};
+use pkcs11::Ctx;
+use simple_asn1::ASN1Block::{BitString, ObjectIdentifier, OctetString, Sequence};
+use simple_asn1::{from_der, oid, to_der, ASN1DecodeErr, ASN1EncodeErr, OID};
+use std::path::Path;
+use std::ptr;
+use thiserror::Error;
+
+type KeyIdVec = Vec<u8>;
+type KeyId = [u8];
+type DerPublicKeyVec = Vec<u8>;
+
+/// Type alias for a sha256 result (ie. a u256).
+type Sha256Hash = [u8; 32];
+
+// We expect the parameters to be curve secp256r1.  This is the base127 encoded form:
+const EXPECTED_EC_PARAMS: &[u8; 10] = b"\x06\x08\x2a\x86\x48\xce\x3d\x03\x01\x07";
+
+/// An error happened related to a HardwareIdentity.
+#[derive(Error, Debug)]
+pub enum HardwareIdentityError {
+    #[error(transparent)]
+    PKCS11(#[from] pkcs11::errors::Error),
+
+    #[error("ASN decode error {0}")]
+    ASN1Decode(ASN1DecodeErr),
+
+    #[error(transparent)]
+    ASN1Encode(#[from] ASN1EncodeErr),
+
+    #[error(transparent)]
+    KeyIdDecode(#[from] hex::FromHexError),
+
+    #[error("Key not found")]
+    KeyNotFound,
+
+    #[error("Unexpected key type {0}")]
+    UnexpectedKeyType(CK_KEY_TYPE),
+
+    #[error("Expected EcPoint to be an OctetString")]
+    ExpectedEcPointOctetString,
+
+    #[error("EcPoint is empty")]
+    EcPointEmpty,
+
+    #[error("Attribute with type={0} not found")]
+    AttributeNotFound(CK_ATTRIBUTE_TYPE),
+
+    #[error("Invalid EcParams.  Expected prime256v1 {:02x?}, actual is {:02x?}", .expected, .actual)]
+    InvalidEcParams { expected: Vec<u8>, actual: Vec<u8> },
+}
+
+/// An identity based on an HSM
+pub struct HardwareIdentity {
+    key_id: KeyIdVec,
+    ctx: Ctx,
+    session_handle: CK_SESSION_HANDLE,
+    logged_in: bool,
+    public_key: DerPublicKeyVec,
+}
+
+impl HardwareIdentity {
+    /// Create an identity using a specific key on an HSM.
+    /// The filename will be something like /usr/local/lib/opensc-pkcs11.s
+    /// The key_id must refer to a ECDSA key with parameters prime256v1 (secp256r1)
+    /// The key must already have been created.  You can create one with pkcs11-tool:
+    /// $ pkcs11-tool -k --slot $SLOT -d $KEY_ID --key-type EC:prime256v1 --pin $PIN
+    pub fn new<P>(
+        filename: P,
```

Comment:
`filename` probably gives the wrong expectation here. What about something like `pkcs11_lib_path`?

Reply:
Sounds good